crates/types/src/string.rs

```diff
@@ -201,16 +180,154 @@ impl StringBuilder {
         debug_assert!(self.inner.len < self.cap);
     }
 
+    /// Initialize [`StringBuilder`] with capacity.
+    pub fn with_capacity(cap: usize) -> Self {
+        if cap == 0 {
+            return Self::new();
+        }
+        let real_cap = cap + 1;
+        let ptr = unsafe { libc::malloc(real_cap) };
+        if ptr.is_null() {
+            unable_to_alloc_memory();
+        }
+        Self {
+            inner: String { len: 0, data: ptr as *mut ffi::c_char },
+            cap: real_cap,
+        }
+    }
+
+    /// Reserve space for N more bytes.
+    ///
+    /// Does not allocate if enough space is available.
+    pub fn reserve(&mut self, cap: usize) {
+        if cap == 0 {
+            return;
+        }
+        let remaining = self.remaining_capacity();
+        if remaining >= cap {
+            return;
+        }
+        let needed_space =
+            if self.inner.data.is_null() { cap + 1 } else { cap - remaining };
+        let new_capacity = (self.cap + needed_space)
+            .checked_next_power_of_two()
+            // in the rare case that the next power of two overflows we should allocate the minimal
+            // amount to avoid OOM problems
+            .unwrap_or(self.cap + needed_space);
+        unsafe { self.reserve_raw(new_capacity) };
+        self.cap = new_capacity;
+    }
+
+    /// Reserve space for exactly N more bytes.
+    ///
+    /// Does not allocate if enough space is available.
+    pub fn reserve_exact(&mut self, cap: usize) {
+        if cap == 0 {
+            return;
+        }
+        let remaining = self.remaining_capacity();
+        if remaining > cap {
+            return;
+        }
+        let needed_space =
+            if self.inner.data.is_null() { cap + 1 } else { cap - remaining };
+        let new_capacity = self.cap + needed_space;
+        unsafe { self.reserve_raw(new_capacity) };
+        self.cap = new_capacity;
+    }
+
+    /// Reserve at least "by" bytes.
+    ///
+    /// This is marked unsafe as the allocation size should never be zero for portability reasons.
+    /// A check isn't included as all call sites should already handle the case.
+    unsafe fn reserve_raw(&mut self, by: usize) {
```

Comment:
`by` is not very descriptive (I'm not even sure what it stands for). Maybe `new_capacity`?